src/Content/Smilies.php

```diff
@@ -152,6 +153,129 @@ public static function getList(): array
+	 *    'icons' => icon url or an empty string
+	 *    'norms' => normalized shortcut
+	 */
+	public static function getNormalizedList(): array
+	{
+		$smilies = self::getList();
+		$norms = [];
+		$icons = $smilies['icons'];
+		foreach ($smilies['texts'] as $i => $shortcode) {
+			// Extract urls
+			$icon = $icons[$i];
+			if (preg_match('/src="(.+?)"/', $icon, $match)) {
+				$icon = $match[1];
+			} else {
+				$icon = '';
+			}
+			$icons[$i] = $icon;
+
+			// Normalize name
+			$norm = preg_replace('/[\s\-:#~]/', '', $shortcode);
+			if (ctype_alnum($norm)) {
+				$norms[] = $norm;
+			} elseif (preg_match('#/smiley-(\w+)\.gif#', $icon, $match)) {
+				$norms[] = $match[1];
+			} else {
+				$norms[] = 'smiley' . $i;
+			}
+		}
+		$smilies['norms'] = $norms;
+		return $smilies;
+	}
+
+	/**
+	 * Finds all used smilies (denoted by quoting colons like :heart:) in the provided text and normalizes their usages.
+	 *
+	 * @param string $text that might contain smiley usages
+	 * @return array with smilie codes (colon included) as the keys, their image urls as values;
+	 *               the normalized string is put under the '' (empty string) key
+	 */
+	public static function extractUsedSmilies(string $text): array
+	{
+		$emojis = [];
+
+		$emojis[''] = BBCode::performWithEscapedTags($text, ['code'], function ($text) use (&$emojis) {
+			return BBCode::performWithEscapedTags($text, ['noparse', 'nobb', 'pre'], function ($text) use (&$emojis) {
+				if (strpos($text, '[nosmile]') !== false || self::noSmilies()) {
+					return $text;
+				}
+				$smilies = self::getNormalizedList();
+				$normalized = array_combine($smilies['texts'], $smilies['norms']);
+				return self::performForEachWordMatch(
+					array_combine($smilies['texts'], $smilies['icons']),
+					$text,
+					function (string $name, string $image) use($normalized, &$emojis) {
+						$name = $normalized[$name];
+						if (preg_match('/src="(.+?)"/', $image, $match)) {
+							$image = $match[1];
+							$emojis[$name] = $image;
+						}
+						return ':' . $name . ':';
+					},
+				);
+			});
+		});
+
+		return $emojis;
+	}
+
+	/**
+	 * Similar to strtr but matches only whole words and replaces texts with $callback.
+	 *
+	 * @param array $words
+	 * @param string $subject
+	 * @param callable $callback ($offset, $value)
+	 * @return string
+	 */
+	private static function performForEachWordMatch(array $words, string $subject, callable $callback): string
+	{
+		$offset = 0;
+		$result = '';
+		$processed = 0;
+		// Learned from PHP's strtr implementation
+		// Should probably improve performance once JIT-compiled
+		$length_bitset = 0;
+		$ord_bitset = 0;
+		foreach ($words as $word => $_) {
+			$length = strlen($word);
+			if ($length <= 31) {
+				$length_bitset |= 1 << $length;
+			}
+			$ord = ord($word);
+			$ord_bitset |= 1 << ($ord & 31);
+		}
+
+		while ($offset < strlen($subject) && preg_match('/\s+?(?=\S|$)/', $subject, $matches, PREG_OFFSET_CAPTURE, $offset)) {
```

Comment:
Words can be delimited by more than whitespace. Punctuation too, for example. is this intended?

Reply:
No, I am still unsure about what qualifies as a smiley... It seems to me that smilies can include literally any character and the only "safe" delimiters are whitespaces. (Or are they? Can smilies contain whitespaces?) For example, treating the following segments as smilies seems a bit strange to me:

- `O.o.O` -> `[emoji=O.o].O`
- `:-;;` -> `[emoji=:-;];`
- `":-"` -> `"[emoji=:-"]`

But these character combinations are rare enough and the users may do intend them as smilies.
Anyway, I've changed the code to recognize smilies surrounded by either spaces or punctuations. (However, currently the code use `ctype_*` function, which has poor support for Unicode chars like `，` (full-width comma, not recognized by `ctype_punct`) or any CJK characters.)

Reply:
I'm with you on the confusion, hence my original question. I personally believe the non-Mastodon standard emojis should be abandoned in favor of regular Unicode emojis.